### PR TITLE
Allow multiple unicast host providers

### PR DESCRIPTION
**Pull request by ywelsch**
This PR introduces support for multiple host providers. This allows the settings based hosts resolver to be just be treated as any other `UnicastHostsProvider`. The PR also introduces the notion of a `HostsResolver` so that plugins such as FileBasedDiscovery do not need to create their own thread pool for resolving hosts, making it easier to add new plugins.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/discovery/zen/SettingsBasedHostsProvider.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.transport.TransportAddress;
+import org.elasticsearch.transport.TransportService;
+
+import java.util.List;
+import java.util.function.Function;
+
+import static java.util.Collections.emptyList;
+
+public class SettingsBasedHostsProvider extends AbstractComponent implements UnicastHostsProvider {
```
Comment (tlrx) [inline server/src/main/java/org/elasticsearch/discovery/zen/SettingsBasedHostsProvider.java]: Can you add some documentation please?

File: server/src/main/java/org/elasticsearch/discovery/zen/SettingsBasedHostsProvider.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    private final List<String> configuredHosts;
+
+    private final int limitPortCounts;
+
+    public SettingsBasedHostsProvider(Settings settings, TransportService transportService) {
+        super(settings);
+
+        if (DISCOVERY_ZEN_PING_UNICAST_HOSTS_SETTING.exists(settings)) {
+            configuredHosts = DISCOVERY_ZEN_PING_UNICAST_HOSTS_SETTING.get(settings);
+            // we only limit to 1 addresses, makes no sense to ping 100 ports
```
Comment (tlrx) [inline server/src/main/java/org/elasticsearch/discovery/zen/SettingsBasedHostsProvider.java]: Nit: addresses -> address

File: server/src/main/java/org/elasticsearch/discovery/zen/UnicastHostsProvider.java [hunk 3]
```diff
@@ -31,5 +31,10 @@
     /**
      * Builds the dynamic list of unicast hosts to be used for unicast discovery.
      */
-    List<TransportAddress> buildDynamicHosts();
+    List<TransportAddress> buildDynamicHosts(HostsResolver hostsResolver);
+
+    interface HostsResolver {
```
Comment (tlrx) [inline server/src/main/java/org/elasticsearch/discovery/zen/UnicastHostsProvider.java]: Can you add a bit of documentation please?

File: server/src/main/java/org/elasticsearch/discovery/zen/UnicastZenPing.java [hunk 4]
```diff
@@ -205,8 +183,13 @@ public UnicastZenPing(Settings settings, ThreadPool threadPool, TransportService
...
                 .map(hn -> (Callable<TransportAddress[]>) () -> transportService.addressesFromString(hn, limitPortCounts))
                 .collect(Collectors.toList());
-        final List<Future<TransportAddress[]>> futures =
-            executorService.invokeAll(callables, resolveTimeout.nanos(), TimeUnit.NANOSECONDS);
+        final List<Future<TransportAddress[]>> futures;
+        try {
+            futures = executorService.invokeAll(callables, resolveTimeout.nanos(), TimeUnit.NANOSECONDS);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            return Collections.emptyList();
```
Comment (tlrx) [inline server/src/main/java/org/elasticsearch/discovery/zen/UnicastZenPing.java]: :+1:

**Review by tlrx — APPROVED** [review]
LGTM, I like it! I left some minor comments around documentation.

**Comment by ywelsch** [comment]
Thanks @tlrx

**Pull request merged**